### ISSUE TITLE
relative positions of key points of shape-shifting animals in yolo

### ISSUE THREAD

**Issue by sronilsson** [ultralytics/ultralytics#21312]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi,
I am using YOLO for pose-estimation of mice and was getting some wonky data relative to pose-estimation tools specialized for my use case.

https://github.com/user-attachments/assets/2134d34c-6197-49be-beb0-1b5d90ab8f68

The left is inference from YOLO and in the right is inference from [SLEAP]( https://github.com/talmolab/sleap). 

The positioning in YOLO is slightly off (relative to ground truth/SLEAP), I appreciate that may be some accuracy tradeoff for latency. 

However, in YOLO, the relative distance between the body-parts appears to be more fixed – such as when the nose is determined to be N pixels from the left/right ears, that distance remains relatively stable in inference even though the distance is decreases/increases, as in the clip below from YOLO.

https://github.com/user-attachments/assets/19a517e3-bd8a-4a86-b5ef-7e20f3cfa0ac

Mice a pretty good shape shifters so do anyone know where this comes from and how I can get rid of it? Cheers!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @sronilsson, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users, where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (such as pose estimation of shape-shifting animals), please provide as much information as possible, including dataset image examples, training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response. An Ultralytics engineer will also assist you here soon. Thank you for your patience and for sharing your use case with us! 🐭✨

**Comment by Y-T-G** [comment]
Can you post your dataset YAML?

**Comment by sronilsson** [comment]
```
task: pose
mode: train
model: D:\yolo_weights\yolo11n-pose.pt
data: D:\cvat_annotations\yolo_07032025\map.yaml
epochs: 1500
time: null
patience: 100
batch: 32
imgsz: 640
save: true
save_period: -1
cache: false
device: '0'
workers: 8
project: D:\cvat_annotations\yolo_07032025\mdl
name: train
exist_ok: false
pretrained: true
optimizer: auto
verbose: true
seed: 0
deterministic: true
single_cls: false
rect: false
cos_lr: false
close_mosaic: 10
resume: false
amp: true
fraction: 1.0
profile: false
freeze: null
multi_scale: false
overlap_mask: true
mask_ratio: 4
dropout: 0.0
val: true
split: val
save_json: false
conf: null
iou: 0.7
max_det: 300
half: false
dnn: false
plots: true
source: null
vid_stride: 1
stream_buffer: false
visualize: false
augment: false
agnostic_nms: false
classes: null
retina_masks: false
embed: null
show: false
save_frames: false
save_txt: false
save_conf: false
save_crop: false
show_labels: true
show_conf: true
show_boxes: true
line_width: null
format: torchscript
keras: false
optimize: false
int8: false
dynamic: false
simplify: true
opset: null
workspace: null
nms: false
lr0: 0.01
lrf: 0.01
momentum: 0.937
weight_decay: 0.0005
warmup_epochs: 3.0
warmup_momentum: 0.8
warmup_bias_lr: 0.1
box: 7.5
cls: 0.5
dfl: 1.5
pose: 12.0
kobj: 1.0
nbs: 64
hsv_h: 0.015
hsv_s: 0.7
hsv_v: 0.4
degrees: 0.0
translate: 0.1
scale: 0.5
shear: 0.0
perspective: 0.0
flipud: 0.0
fliplr: 0.5
bgr: 0.0
mosaic: 1.0
mixup: 0.0
cutmix: 0.0
copy_paste: 0.0
copy_paste_mode: flip
auto_augment: randaugment
erasing: 0.4
cfg: null
tracker: botsort.yaml
save_dir: D:\cvat_annotations\yolo_07032025\mdl\train
```


Thanks @Y-T-G - is the above what you are looking for?

**Comment by Y-T-G** [comment]
The content of map.yaml

**Comment by sronilsson** [comment]
@Y-T-G 

This one ?

```
path: D:\cvat_annotations\yolo_07032025
train: images\train
val: images\val
kpt_shape: [9, 3]
flip_idx: [0, 2, 1, 3, 5, 4, 6, 7, 8]
names:
  0: mouse skel
```

**Comment by Y-T-G** [comment]
Is the issue present in the training batch plots? The one inside the runs folder with filename `train_batchX_labels.jpg`?

**Comment by sronilsson** [comment]
Thanks @Y-T-G  no it's not obvious in the plots or present in my annotations.. like below. 

![Image](https://github.com/user-attachments/assets/61dccc92-5325-4010-bbb4-0ac5c8939655)

...  in some of these images, like index `1, 2` and `2, 1`  in the mosaic I can  see the the bboxes plotted but not the keypoints, even though the training label txt has all keypoints marked as present. but I assumed that was unrelated to the current issue.

s27_whisking_hq_01199.txt
```
0 0.632859375 0.22823611111111108 0.617703125 0.3357777777777778  0.30703125 0.08472222222222223 0.0 0.38828125 0.24027777777777778 2.0 0.3234375 0.12777777777777777 2.0 0.5609375 0.24166666666666667 2.0 0.540625 0.0763888888888889 2.0 0.43046875 0.0 0.0 0.6484375 0.059722222222222225 2.0 0.8328125 0.13472222222222222 2.0 0.94140625 0.3958333333333333 2.0
```

**Comment by Y-T-G** [comment]
Were the boxes auto generated? You should add some margin to it so that the points are not at the border. Ideally, the boxes need to correctly cover the object's contours.

**Comment by sronilsson** [comment]
Cheers, that makes sense. I created the boxes, as the minimum axis-aligned box encompassing all my annotated points. Any recommendation how much padding there should be - like is it enough with a pixel?

Do you think the lack of box padding is producing the issues, with poorer pose and  the issue with inferred key-point distances being  kinda locked between frames?

**Comment by Y-T-G** [comment]
The correct is to cover the subject with the box fully. So padding is a hack in this case and probably causes issues in training due to inconsistency between labels and features.

You can try making all the boxes 10% larger.

**Comment by sronilsson** [comment]
Got it thanks @Y-T-G - i will try that, close this for now, and reopen if issue persist 👍

**Comment by sronilsson** [comment]
@Y-T-G after some dabbling… it doesn’t appear to be related to the bbox size/padding. Rather it appears related to the `imgsz` passed during fitting vs the `imgsz` passed during inference. For example, in the video below, I use a pose estimation model fitted using **imgsz=640**, for inference using **imgsz=320, 640, 1280**. The far best inference is when imgsz is the same as during fitting. Is this to be expected, or do you see any other params that could be playing?

https://github.com/user-attachments/assets/e9289871-e281-4bd5-bc06-d6bbbb4e15dd

**Comment by Y-T-G** [comment]
Yes, that's expected. The model performs best at the `imgsz` it was trained at. Unless the objects themselves are changing sizes to compensate for the difference. For example, running with `imgsz=320`, but the objects are also 2x larger relative to the whole image compared to its size at `imgsz=640` (which was used for training). Or at `imgsz=1280` when the objects are 2x smaller.
